src/java.base/share/classes/java/util/Currency.java

```diff
@@ -432,61 +433,78 @@ public static Currency getInstance(Locale locale) {
-                    for (char c2 = 'A'; c2 <= 'Z'; c2 ++) {
-                        int tableEntry = getMainTableEntry(c1, c2);
-                        if ((tableEntry & COUNTRY_TYPE_MASK) == SIMPLE_CASE_COUNTRY_MASK
-                             && tableEntry != INVALID_COUNTRY_ENTRY) {
-                            char finalChar = (char) ((tableEntry & SIMPLE_CASE_COUNTRY_FINAL_CHAR_MASK) + 'A');
-                            int defaultFractionDigits = (tableEntry & SIMPLE_CASE_COUNTRY_DEFAULT_DIGITS_MASK) >> SIMPLE_CASE_COUNTRY_DEFAULT_DIGITS_SHIFT;
-                            int numericCode = (tableEntry & NUMERIC_CODE_MASK) >> NUMERIC_CODE_SHIFT;
-                            StringBuilder sb = new StringBuilder();
-                            sb.append(c1);
-                            sb.append(c2);
-                            sb.append(finalChar);
-                            available.add(getInstance(sb.toString(), defaultFractionDigits, numericCode));
-                        } else if ((tableEntry & COUNTRY_TYPE_MASK) == SPECIAL_CASE_COUNTRY_MASK
-                                && tableEntry != INVALID_COUNTRY_ENTRY
-                                && tableEntry != COUNTRY_WITHOUT_CURRENCY_ENTRY) {
-                            int index = SpecialCaseEntry.toIndex(tableEntry);
-                            SpecialCaseEntry scEntry = specialCasesList.get(index);
-
-                            if (scEntry.cutOverTime == Long.MAX_VALUE
-                                    || System.currentTimeMillis() < scEntry.cutOverTime) {
-                                available.add(getInstance(scEntry.oldCurrency,
-                                        scEntry.oldCurrencyFraction,
-                                        scEntry.oldCurrencyNumericCode));
-                            } else {
-                                available.add(getInstance(scEntry.newCurrency,
-                                        scEntry.newCurrencyFraction,
-                                        scEntry.newCurrencyNumericCode));
-                            }
+        initAvailableCurrencies();
+        return new HashSet<>(available);
+    }
+
+    /**
+     * {@return a stream of available currencies} The returned stream of currencies
+     * contains all the available currencies, which may include currencies
+     * that represent obsolete ISO 4217 codes. If there is no currency
+     * available in the runtime, the returned stream is empty.
+     *
+     * @implNote Unlike {@link #getAvailableCurrencies()}, this method does
+     * not create a defensive copy of the {@code Currency} set.
+     * @since 25
+     */
+    public static Stream<Currency> availableCurrencies() {
+        initAvailableCurrencies();
+        return available.stream();
+    }
+
+    // Initialize the set of available currencies if needed
+    private static synchronized void initAvailableCurrencies() {
+        if (available == null) {
+            available = new HashSet<>(256);
```

Comment:
Not related to your change, but this should be `HashSet.newHashSet(256)`. In fact 256*0.75 is 192, which is smaller than the current number of currencies (230, as of JDK24)

Reply:
Fixed in https://github.com/openjdk/jdk/pull/23165/commits/c94ba9447cf3d24ba6e4162890a9be5d611dbfa7